README
$Id: README,v 1.15 2006/12/11 21:50:33 queinnec Exp $

(((((((((((((((((((((((((((((((( L i S P ))))))))))))))))))))))))))))))))
 This file is part of the files that accompany the book:
     LISP Implantation Semantique Programmation (InterEditions, France)
 By Christian Queinnec <Christian.Queinnec@INRIA.fr>
 Newest version may be retrieved from:
      http://www-spi.lip6.fr:~queinnec/Books/LiSP-2ndEdition.tgz
(((((((((((((((((((((((((((((((( L i S P ))))))))))))))))))))))))))))))))

                           Programs of
                       Lisp in Small Pieces

                <Christian.Queinnec@polytechnique.fr>


These are the source files that accompany the second French edition of
the book "Principes d'implantation de Scheme et Lisp" by Christian
Queinnec. These are roughly the same programs that were bundled with
the English edition and no effort was made to make them run with
modern Scheme systems.

Although probably not bug-free, these programs are made available as
they are. No warranty is made about these programs or their
performance. They are distributed in the hope that they will be useful
and help readers to experiment with the concepts of the associated
book.  Use and copying of this software and preparation of derivative
works based upon this software are permitted, so long as the following
conditions are met:
      o credit to the author is acknowledged according to current
        academic practices
      o no fees or compensation are charged for use, copies, or
        access to this software
      o this copyright notice is included intact.

Remember that these programs were used to build a pedagogical book. In
many places, the algorithms are not optimal and rather naive. Better
algorithms may be found in other places but are there to show
variants.  These programs contain a lot of material (more than 13
different interpreters or compilers) written with many different
styles (naive, object-oriented, continuation-passing, closure-passing,
byte-code compilation, compilation towards C, etc). The intention is
to show again and again the various facets of the semantics of
Lisp-like languages.

____________________________________________________________________________
*** SKETCH OF THE TABLE OF CONTENTS

Chapter 1: The basic interpreter (eval expression environment).

Chapter 2: Name spaces and Recursion
        Some new interpreters introducing lexical and/or dynamic binding,
        single or multiple namespaces, various brands of global environments.

Chapter 3: Escape, Continuation
        A bunch of control operators (catch/throw, block/return-from,
        unwind-protect, call/cc) all explained through an interpreter
        (eval exp env continuation) written in OO style.

Chapter 4: Side-effect
        Assignment, data mutation and equality explained through another
        interpreter (eval exp env cont memory) written with lambdas only.

Chapter 5: Denotational Semantics
        A simple currying of the previous interpreter which now appears as
        ((eval exp) env cont mem). Numerous variants such as dynamic
        binding, global environments.

Chapter 6: Fast interpretation
        Precompile expressions to speed up interpretation. Introduce
        runtime representations of environment and continuations through
        various interpreters such as
                ((eval exp lexical.env) runtime.env continuation),
                ((eval exp lexical.env) runtime.env) and finally
                ((eval exp lexical.env)).

Chapter 7: ByteCode compilation
        Compile expressions into byte-code instructions; invent registers,
        PC, stack and the like with still another compiler/interpreter
                (bytecode-run (bytecode-compile exp lexical.env))

Chapter 8: Eval and other reflective features.
        A whole chapter on eval, its implementation and cost, in all the
        previous interpreters/compilers. Also introduce first class
        environments and a complete reflective interpreter.

Chapter 9: Macros
        Everything you want to know on macros, macroexpansion, hygien etc.
        in relation with (separate) compilation and interpretation.

Chapter 10: Compilation towards C
        Another complete compiler from Scheme to C and its runtime in C.

Chapter 11: An Object System (Meroonet)
        The innards of an Object system (based on Meroon) in Scheme.

  There are also corrected exercices, a bibliography and an index.

____________________________________________________________________________
*** INSTALLATION

These files were tested with a Scheme interpreter augmented with
        a test-suite driver (tester.scm),
        the define-syntax and define-abbreviation macros (using
                Dybvig's syntax-case package),
        and an object system: Meroonet (meroonet.scm).

Bigloo, Scheme->C, Gambit, Elk or SCM can be used. The first three are
better since a specialized interpreter may be built that contains a
compiled Meroonet and compiled hygienic macros. To obtain such an
interpreter you'll have to examine the Makefile of the distribution,
mainly setup the SCHEME variable then, run `make' to regenerate this
interpreter that will appear in the o/${HOSTTYPE} directory.

Of course, on Mac, you cannot rebuild things so easily. With MacGambit
you must load or compile the gambit/book.scm file.

____________________________________________________________________________
*** CONVENTIONS

Source files are named after the chapter to which they belong. The
name of the file is built as chap<number><letters>.scm where <number>
is the number of the chapter and <letters> differentiate variants or
parts. For many of these files, there exist specific test suites with
a similar name but with a .tst extension. These files have some
dependencies that are visible in the Imakefile that shows how to load
them and test them. Most of these files have an associated test entry
in the Imakefile named test.chap<number><letters>. Test suites may
usually be run with a Scheme function named test-{scheme,chap}<number>
while interpreters may be started with a scheme<number> function.

Examples:

You want to use the OO-style interpreter of chapter 3.  Scan Makefile
for "test.chap3", you will find a line saying that chap3f contains
such an interpreter. To try this interpreter, you have to load files
src/chap3f.scm, src/chap3g.scm and src/chap3h.scm. Then, given that
tests are started with test-scheme3f, the interpreter is started with
(scheme3f). This can be confirmed by looking at the end of file
chap3f.scm. And you're done!

You now want to use the Bytecode compiler. Scan Makefile or the rest
of this file to find an occurence of `bytecode', this yields chap7d.
Look now for the `test.chap7d' entry in Makefile to discover that you
have to load files src/chap6d.scm and src/chap7d.scm. Look at the end
of src/chap7d how to run this new system: (scheme7d) is the answer.

If you want to load the compiler from Scheme towards C, then look in
the index below. You'll find that this compiler is defined in
chap10e.scm then look in the Imakefile file for the test.chap10e
entry. You'll find there that you have to load chap10a, chap10c,
chap10g, chap10e, chap10h, chap10f to obtain it. Then looking at these
files, you'll figure out how to run the compiler (Hint: look at the
show-C-expression function from the chap10f.scm file). Of course, more
documentation is given in the associated book!


____________________________________________________________________________
*** REMARKS

Bug descriptions, use reports, comments or suggestions are welcome
whether on these programs or on the book.

Send them to:                    or to:

ChangeLog
2006-12-04  Christian Queinnec  <Christian.Queinnec@lip6.fr>

	* LiSP-2ndEdition.tgz : Shortened lines in some files to cope with
	the new French edition that is printed on a slightly less larger
	paper.

	* fixed some years ago a bug in cps-begin visible in the first
	Edition (French or English): the order of evaluation was wrong.
	Thanks to Luc Moreau do have discovered it.

Mon Jan 29 20:49:07 1996  Christian Queinnec  <queinnec@blaye.inria.fr>

        * Book.mkf : Written customization for Gambit and Elk so you can
        run the programs of the book with them. Still a problem with
        syntax-case with Gambit and don't use -l in Elk since you loose
        top-level-environment at that time (so you have to load
        "elk/book.elk" by hand).

        * Many files: Changed slightly many files for names, presentation,
        comments etc to ease the English translation.

Thu Oct  6 11:00:50 1994  Christian Queinnec  <queinnec@blaye.inria.fr>

        * src/chap4a.scm : globalized the local `shared' variable so it
        can be reset to () in file-test. This fix allows to remove lots of
        annoying quotations that appeared in scheme.tst.

Fri Sep  2 14:53:49 1994  Christian Queinnec  (queinnec@blaye.inria.fr)

        * src/chap9e.scm ((->Scheme (e Function) r)): contains a call to
        gensym with an argument: removed that argument.
        Also corrected src/chap9c.tst which contains a bugged test.

Wed Aug 31 11:12:17 1994  Christian Queinnec  (queinnec@blaye.inria.fr)

        * src/chap6f.scm ((generate-user-global-variables out)): global
        variables are printed in uppercase since rt.h supposed it and scm
        prints it in lowercase by default.

        * src/chap8k.scm (reflisp): the macroexpansion (monitor and
        the-environment are the only macros) is now done by hand.

        * src/chap7g.scm ((test-scheme7g file)): To protect the si
        directory that is used to regenerate the book with hand-improved
        programs, change si/ into tmp.si/ for tests. Also change Test.mkf.

Tue Aug 30 19:50:56 1994  Christian Queinnec  (queinnec@blaye.inria.fr)

        * src/chap6f.scm ((get-reference value)): It does not seem
        mandatory in Scheme that symbol->string gives eq results for a
        singe symbol !? So changed eq? to equal? in that function.

        * Test.mkf (GRAND_TEST_FLAGS): Added this entry to check all tests
        as well as a perl script to inspect the results.

        * src/chap6e.scm ((PREDEFINED i)): added that instruction which
        was missing.

Mon Aug 29 11:22:12 1994  Christian Queinnec  (queinnec@blaye.inria.fr)

        * src/chap10m.scm: Made a file of letify (changed also its name
        from letify! to letify since there are no longer side-effects) to
        ease testing it. Removed the patch from src/chap10c.scm (the
        method on lift-procedure!).

Sun Aug 28 11:12:06 1994  Christian Queinnec  (queinnec@blaye.inria.fr)

        * src/chap10c.scm ((lift-procedures! (o Free-Reference) flatfun
        vars)): Added that due to a bug wrt CPS. Bug triggered by the `if'
        form in (lambda (l c)
                  (if c l #t)
                  (lambda (max min) l c ) )
        which duplicates the continuation and lift! walks two times this
        continuation.

        * src/chap10k.scm ((test-scheme10k file)): changed equal to
        naive-match since test suite chap10k.tst needs it.

        * bigloo/compapp.scm (compile-bigloo-application): force the order
        of expansion in case of multiple files. Added *verbose* flag.

        * src/chap10h.scm: commented definitions of generate-declarations
        and generate-prototypes which were obsolete (still using
        Functional-Description-Cname).

        * src/chap10f.scm ((compiler-entry-point options)): Added this

src/c/rt.h
/* $Id: rt.h,v 4.0 1995/07/10 06:52:27 queinnec Exp $ */

/* (((((((((((((((((((((((((((((((( L i S P ))))))))))))))))))))))))))))))))
 * This file is part of the files that accompany the book:
 *     LISP Implantation Semantique Programmation (InterEditions, France)
 * By Christian Queinnec <Christian.Queinnec@INRIA.fr>
 * Newest version may be retrieved from:
 *   (IP 128.93.2.54) ftp.inria.fr:INRIA/Projects/icsla/Books/LiSP*.tar.gz
 * Check the README file before using this file.
 *(((((((((((((((((((((((((((((((( L i S P ))))))))))))))))))))))))))))))))
 */

/* The C header describing the run-time of chap6f
 * Scheme objects representation.
 * All objects are boxed for simplicity.			        */

/* Prevent double inclusion */
#ifndef RTLIB_SCHEME_H
#define RTLIB_SCHEME_H

/* TEMP pour l'instant                        TEMP */
#define NO_GC

/* Compilation products only need to include this file which in turn
 * include what is needed. This simplifies code generation.		*/
#include <stdio.h>
#include <sys/file.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <setjmp.h>
#include <stdlib.h>

/***********************************************************************
 * General machine dependencies:
 * SonyNews ignores size_t so we redefine some primitives that return
 * size_t values. We thus define and use SIZE_T for all natural numbers
 * that may be used to compute on addresses.				*/
#ifdef sony_news
typedef unsigned int SIZE_T ;
#else
typedef size_t SIZE_T ;
#endif

#define MALLOC(x) ((void*) malloc((SIZE_T) x))
#define SIZEOF(x) ((SIZE_T) sizeof(x))
#define STRLEN(x) ((SIZE_T) strlen(x))

/* this macro tries to concatenate two tokens into a single one.
 * The first definition only works in ANSI-C (and gcc)
 * The second works on most of other cc.				*/
#ifdef __STDC__
#define SCM_CONC(name1,name2) name1##name2
#else
#define SCM_CONC(name1,name2) name1/**/name2
#endif

/* This macro computes the address which separates two ordered addresses.
 * Better that the result is word-aligned since it will be used as the
 * starting address of a heap.						*/
#define SCM_mid_address(start,end) \
  ((char*) start + (((((char*) end) - ((char*) start))/2) & ~3))

/***********************************************************************
 * The library is compiled with respect to the following constants.
 * If you want to change them you have to recompile the library for them
 * to take effect.							*/

/* This constant controls whether some dynamic checks are performed or
 * not. For example whether the last argument of apply is a real list.
 *                  1 means true, 0 means false.
 * If you set it to false, the associated tests will become dead code.
 * Usual compilers are able to remove it automatically.			*/
#define SCM_CAREFUL 1

/* This constant defines the characters that are predefined and prestored
 * in an array. 256 corresponds to ISO_Latin1 code size and 128 is also
 * possible for old Ascii coding. Smaller values are not advised.
 * NOTE: This constant must be a power of 2 (see SCM_charcodep()).	*/
#define SCM_CHAR_SIZE 256

/* This constant defines the greatest small fixnum which is preallocated.
 * When a fixnum must be allocated, the preallocated corresponding one
 * is returned if it exists. The gain is considerable even for (fib 20).
 * You can set this number to zero if you do not want preallocated
 * fixnums. 								*/
#define SCM_SMALL_FIXNUM_LIMIT 1024

/* This constant defines the maximal number of arguments a function
 * called by apply can take. Actually the interface for functions with
 * large number of variables allows any number of arguments. Due to the
 * actual implementation of apply, there is a limit which might be
 * removed in future releases.
 * 256 is a good choice but you can enlarge it if you consider having
 * functions with more than 256 actual arguments.			*/
#define SCM_MAX_APPLY_ARGUMENTS 256

/* This constant defines the unit (in bytes) of the size of the zones
 * that will be allocated. 						*/
#define SCM_ZONE_SIZE 1024

/* This constant is the ratio of the size of a space to the occupied
 * size. Appel says that 8 is correct ie a stop-and-copy works well
 * when there is 8 more space than the total space used.		*/
#define SCM_GC_EXTENSION_THRESHOLD 8

/***********************************************************************
 * All external names defined by this header file are prefixed by SCM.
 * The C functions that will become the behavior of Scheme functions
 * ie have the correct signature, have a second capitalized name. For
 * example SCM_Cons. Predicates have a P suffix like SCM_EqP.
 * There is a lot of macros or small functions like SCM_cons which
 * implements the effect associated to their name. They exist for internal
 * use and are generally unsafe. Their name (apart the SCM_ prefix) only
 * contains small letters. Scheme global variables are in uppercase
 * and prefixed with SCM_V_. Legal characters of Scheme that are illegal
 * for C are mapped on legal C characters. For instance, pair? becomes
 * SCM_V_PAIRP while symbol->string becomes SCM_V_SYMBOL2STRING.

/***********************************************************************
 * SCM_object (see below) is the type of all Scheme objects
 * SCM_reference is the type of references onto these objects.
 * Usually regular functions just use references everywhere so
 * SCM_reference is aliased to SCM as a shorthand which allows one
 * to believe that only objects are handled.				*/

typedef union SCM_object *SCM_reference ;
typedef SCM_reference    SCM	        ;

/***********************************************************************
 * All allocated objects have a header containing a tag from which
 * their type can be derived. This tag is an unsigned int rather than
 * an enum type since it might be extended dynamically.			*/

typedef unsigned int SCM_TAG ;

#define SCM_OBJECT_TAG		9       /* Never zero ! */
#define SCM_CLASS_TAG		1
#define SCM_GENERIC_TAG		2
#define SCM_FIELD_TAG		3
#define SCM_MONO_FIELD_TAG	4
#define SCM_POLY_FIELD_TAG	5
#define SCM_ANOMALY_TAG		6
#define SCM_MODULE_TAG		7
#define SCM_ROOT_TAG		8
#define SCM_PAIR_TAG		10	/* for pairs			*/
#define SCM_SYMBOL_TAG		11	/* for symbols			*/
#define SCM_EMPTY_TAG		12	/* ()				*/
#define SCM_CLOSURE_TAG		13	/* for closures			*/
#define SCM_PRIMITIVE_TAG	14	/* for primitives 		*/
#define SCM_FIXNUM_TAG		15	/* for small integers 		*/
#define SCM_UNDETERMINED_TAG	16	/* The UFO			*/
#define SCM_TRUE_TAG		17	/* #t				*/
#define SCM_FALSE_TAG		18	/* #f				*/
#define SCM_EOF_TAG		19	/* The end of file		*/
#define SCM_BOX_TAG		20	/* for boxes (mutable variables)*/
#define SCM_STRING_TAG		21	/* for strings			*/
#define SCM_CHAR_TAG		22	/* for characters		*/
#define SCM_PORT_TAG		23	/* for ports (streams)		*/
#define SCM_FLOATNUM_TAG	24	/* for float (timings !:)	*/
#define SCM_STACK_SLICE_TAG	25	/* for stack_slice		*/
#define SCM_CONTINUATION_TAG	26	/* for continuations		*/
#define SCM_PROCESS_TAG		27	/* for processes		*/

#define SCM_FRAME_TAG           30      /* for environment frames       */
#define SCM_FUNCTION_TAG        31      /* for functions                */
#define SCM_NULLENV_TAG         32      /* for end of environments      */
#define SCM_DYNENV_TAG          33      /* for dynamic environment      */

/***********************************************************************
  Scheme objects							*/

typedef struct {
  SCM_TAG tag ;
  SCM link ; /* for a stop and copy GC */
  } SCM_object_header ;

union SCM_object {
  struct SCM_immediate_object {
    SCM_object_header header ;
  } object ;
  struct SCM_fixnum {
    SCM_object_header header ;
    int value ;
  } fixnum ;
  struct SCM_floatnum {
    SCM_object_header header ;
    float value ;
  } floatnum ;
  struct SCM_pair {
    SCM_object_header header ;
    SCM cdr ;
    SCM car ;
  } pair ;
  struct SCM_box {
    SCM_object_header header ;
    SCM content ;
  } box ;
  struct SCM_primitive {
    SCM_object_header header ;
    /* A pointer to a function that returns a Scheme value */
    SCM (*behavior)() ;
    int arity ;
    SCM name ;           /* for debug purposes, primitives have names */
  } primitive ;
  struct SCM_closure {
    SCM_object_header header ;
    SCM (*behavior)() ;
    int arity ;
    SIZE_T size ;
    SCM environment[1] ; /* A poly-field */
  } closure ;
  struct SCM_process {
    SCM_object_header header ;
    SCM next ;          /* doubly linked list */
    SCM previous ;
    SIZE_T size ;
    SCM term[2] ;      /* A poly-field */
  } process ;
  struct SCM_module {
    SCM_object_header header ;
    SIZE_T size ;
    SCM global[1] ;      /* A poly-field */
  } module ;
  struct SCM_root {
    SCM_object_header header ;
    SIZE_T size ;
    SCM module[1] ;      /* A poly-field */
  } root ;
  struct SCM_char {
    SCM_object_header header ;
    char C_char ;
  } character ;
  struct SCM_symbol {
    SCM_object_header header ;
    SCM string ;
  } symbol ;
  struct SCM_port {
    SCM_object_header header ;
    int flag ;            /* O_RDONLY, O_WRONLY ... */
    SCM poken_character ; /* (SCM) NULL or SCM_eof or a Scheme character */
    FILE *file ;
  } port ;
  struct SCM_string {
    SCM_object_header header ;
    SIZE_T size ;
    char *C_string ;
    /* There might be characters here if dynamically created. */
  } string ;
  struct SCM_continuation {
    SCM_object_header header ;
    SCM transformer ;
    SCM below ;
  } continuation ;
  struct SCM_stack_slice {
    SCM_object_header header ;
    SIZE_T stack_slice_length ;
    jmp_buf stack_jmp_buf ;
    char stack_char[1] ; /* A poly-field */
  } stack_slice ;
  struct SCM_frame {
    SCM_object_header header ;
    SCM next ;
    SIZE_T size ;
    SCM value[1] ;       /* A poly-field */
  } frame ;
  struct SCM_function {
    SCM_object_header header ;
    SCM (*behavior)() ;
    SCM environment ;
  } function ;
  struct SCM_dynamic_link {
    SCM_object_header header ;
    SCM next ;
    SCM key ;
    SCM value ;
  } dynamic_link ;
} ;

/***********************************************************************
 * Arity is coded as a positive integer if fixed (ie the arity of cons is 2)
 * and as a negative integer if a minimal arity exists. The arity
 * of list is -1 (ie at least zero argument).				*/

/***********************************************************************
 * Prototypes of library functions or macros that might appear in
 * generated C.								*/

extern SCM SCM_allocate(SCM_TAG tag, SIZE_T size) ;
#define SCM_tag_of(x)        (((x)->object).header).tag
/* this compares pointers to data objects */
#define SCM_eqp(x,y)         ((void*)(x)==(void*)(y))
/* while this compares pointers to C functions */
#define SCM_Cfunctions_eqp(x,y) ((void(*)())(x)==(void(*)())(y))
/* Compare addresses rather than tags.
 * Note: This imposes a unique undef object.				*/
#define SCM_undeterminedp(x) (SCM_eqp(x,SCM_undef))
/* Convert a C boolean value to a Scheme boolean value			*/
#define SCM_2bool(x)         ((x)?SCM_true:SCM_false)

extern SCM SCM_make_fixnum (int value) ;
extern struct SCM_fixnum SCM_small_fixnums[] ;
/* SCM_fixnump (as all other predicates with a trailing p, returns
 * a C boolean rather than a Scheme boolean.				*/
#define SCM_fixnump(x)       (SCM_FIXNUM_TAG==SCM_tag_of(x))
#define SCM_fixnum2int(x)    ((int)(x->fixnum).value)

extern SCM SCM_make_floatnum (float value) ;
#define SCM_floatnump(x)       (SCM_FLOATNUM_TAG==SCM_tag_of(x))
#define SCM_floatnum2float(x)  ((float)(x->floatnum).value)

extern SCM SCM_make_pair (SCM car, SCM cdr) ;
#define SCM_cons(x,y)        SCM_make_pair((x),(y))
#define SCM_car(x)           ((x)->pair).car
#define SCM_cdr(x)           ((x)->pair).cdr
#define SCM_set_car(x,y)     (SCM_car(x) = (y))
#define SCM_set_cdr(x,y)     (SCM_cdr(x) = (y))
#define SCM_pairp(x)         (SCM_PAIR_TAG==SCM_tag_of(x))
/* Compare addresses rather than tags
 * #define SCM_nullp(x)         (SCM_EMPTY_TAG==SCM_tag_of(x))		*/
#define SCM_nullp(x)         (SCM_eqp(x,SCM_empty))
#define SCM_pairp2bool(x)    (SCM_2bool(SCM_pairp(x)))

extern SCM SCM_make_continuation (SCM t, SCM q) ;
#define SCM_continuation2transformer(x) ((x)->continuation).transformer
#define SCM_continuation2below(x)       ((x)->continuation).below

extern SCM SCM_make_box (SCM content) ;
#define SCM_boxp(x)              (SCM_BOX_TAG==SCM_tag_of(x))
#define SCM_box_content(x)       ((x)->box).content
#define SCM_set_box_content(x,y) (SCM_box_content(x) = (y))

extern SCM SCM_make_symbol (SCM string) ;
#define SCM_symbolp(x)       (SCM_SYMBOL_TAG==SCM_tag_of(x))
#define SCM_symbol2string(x) ((x)->symbol).string

extern SCM SCM_make_string (char* C_string) ;
extern SCM SCM_allocate_string (SIZE_T size) ;
#define SCM_stringp(x)        (SCM_STRING_TAG==SCM_tag_of(x))
#define SCM_string2Cstring(x) ((x)->string).C_string
#define SCM_string2size(x)    ((x)->string).size
#define SCM_mutable_stringp(x) \
   ( ((char*)SCM_string2Cstring(x)) == ((char*)SCM_string2Cstring(x))+1 )

extern struct SCM_char SCM_characters[] ;
#define SCM_charp(x)          (SCM_CHAR_TAG==SCM_tag_of(x))
/* returns a C int instead of a SCM int.
 * #define SCM_char2int(x)       ((int)((x)->character).C_char)	*/
#define SCM_int2char(x)       ((SCM) &(SCM_characters[x]))
#define SCM_char2int(x)       (((struct SCM_char*)(x))-SCM_characters)

#define SCM_streamp(x)            (SCM_PORT_TAG==SCM_tag_of(x))
#define SCM_port2stream(x)        (((x)->port).file)
#define SCM_port2poken_char(x)    (((x)->port).poken_character)

extern SCM SCM_make_primitive(SCM (*behavior)(), int arity, SCM name) ;
#define SCM_primitivep(x)         (SCM_PRIMITIVE_TAG==SCM_tag_of(x))
/* Returns a SCM object (generally a string).				*/
#define SCM_primitive2name(x)     (((x)->primitive).name)

extern SCM SCM_make_closure(SCM (*behavior)(), int arity, SIZE_T size) ;
#define SCM_closurep(x)         (SCM_CLOSURE_TAG==SCM_tag_of(x))
#define SCM_closed_value(c,i)   (((c)->closure).environment[i])
#define SCM_set_closed_value(c,i,x) (SCM_closed_value(c,i)=(x))
#define SCM_minimal_arity(arity)  (-(arity+1))
#define SCM_CreatePrimitive(SCM_var,C_function,arity,C_string) \
{ static /*const*/ char *name = C_string ; 			        \
  SCM_var = SCM_make_primitive(C_function,arity,SCM_make_string(name)) ; }

extern SCM SCM_make_root (SIZE_T size) ;
extern SCM SCM_make_module (SIZE_T size) ;

/* This is used in rtd_scheme.c to copy slices of the C stack. */
extern SCM SCM_make_stack_slice (SIZE_T size) ;

/* This is used by the GC to determine if an object is static. These
 * are immutable objects and writers verify that.			*/
#ifdef GC_STOP_AND_COPY
#define SCM_staticp(x)   ((char*)(x) < ((char*) zone_static_end) )
#else
#define SCM_staticp(x)   0
#endif /* GC_STOP_AND_COPY */

/***********************************************************************
 * Some constant values. These constants are roots for the GC and must
 * known, so define them with DefineRoot.				*/

extern SCM SCM_true, 		/* #t */
  SCM_false, 			/* #f */
  SCM_empty, 			/* () */
  SCM_undef, 			/* undefined */
  SCM_eof ;			/* end-of-file */
extern struct SCM_immediate_object SCM_true_object,
  SCM_false_object,
  SCM_empty_object,
  SCM_undef_object ;
extern SCM SCM_stdin, 		/* stdin */
  SCM_stdout, 			/* stdout */
  SCM_stderr ;			/* stderr */
extern SCM SCM_options ;	/* the list of options for the command */
extern void *SCM_Cstack_bottom ;	/* the address of the bottom
                                           of the stack (useful to copy
                                           stack slices). */

typedef enum SCM_stack_growing_type { UP=1, DOWN } SCM_stack_growing_type ;
extern SCM_stack_growing_type SCM_stack_grows ; /* the way the stack grows */

typedef enum SCM_gc_state { GC_NEEDED=1, GC_NOT_NEEDED } SCM_gc_state ;
extern SCM_gc_state SCM_gc_neededp(SIZE_T room) ;
extern void SCM_gc (unsigned int number, SCM roots[]) ;
/* This structure will reference all the module objects of the application.
 * These in turn will reference all global variables.			*/
extern SCM SCM_roots ;

#define SCM_truep(x)    (SCM_TRUE_TAG==SCM_tag_of(x))
#define SCM_falsep(x)   (SCM_FALSE_TAG==SCM_tag_of(x))
#define SCM_emptyp(x)   (SCM_EMPTY_TAG==SCM_tag_of(x))
#define SCM_undefp(x)   (SCM_UNDETERMINED_TAG==SCM_tag_of(x))
#define SCM_eofp(x)     (SCM_EOF_TAG==(SCM_tag_of(x)))

/* Some global functions or variables					*/

extern SCM SCM_PairP (SCM x) ;
extern SCM SCM_NullP (SCM x) ;
extern SCM SCM_FixnumP (SCM x) ;
extern SCM SCM_SymbolP (SCM x) ;
extern SCM SCM_StringP (SCM x) ;
extern SCM SCM_CharP (SCM x) ;
extern SCM SCM_EofP (SCM x) ;
extern SCM SCM_ProcedureP (SCM x) ;
extern SCM SCM_EqP (SCM x, SCM y) ;
extern SCM SCM_Cons (SCM x, SCM y) ;
extern SCM SCM_Car (SCM x) ;
extern SCM SCM_Cdr (SCM x) ;
extern SCM SCM_Set_Car (SCM x, SCM y) ;
extern SCM SCM_Set_Cdr (SCM x, SCM y) ;
extern SCM SCM_Plus (SCM x, SCM y) ;
extern SCM SCM_Minus (SCM x, SCM y) ;
extern SCM SCM_Times (SCM x, SCM y) ;
extern SCM SCM_Quotient (SCM x, SCM y) ;
extern SCM SCM_Remainder (SCM x, SCM y) ;
extern SCM SCM_GtP (SCM x, SCM y) ;
extern SCM SCM_LtP (SCM x, SCM y) ;
extern SCM SCM_GeP (SCM x, SCM y) ;
extern SCM SCM_LeP (SCM x, SCM y) ;
extern SCM SCM_EqnP (SCM x, SCM y) ;
extern SCM SCM_Basic_Write (SCM x) ;
extern SCM SCM_Integer2Char (SCM x) ;
extern SCM SCM_Char2Integer (SCM x) ;
extern SCM SCM_Symbol2String (SCM x) ;
extern SCM SCM_Make_Symbol (SCM x) ;
extern SCM SCM_Exit (SCM x) ;
extern SCM SCM_Get_Universal_Time () ;
extern SCM SCM_Float_Plus (SCM x, SCM y) ;
extern SCM SCM_Float_Minus (SCM x, SCM y) ;
extern SCM SCM_Float_Multiply (SCM x, SCM y) ;
extern SCM SCM_Float_Divide (SCM x, SCM y) ;
extern SCM SCM_Float_GreaterP (SCM x, SCM y) ;
extern SCM SCM_Float_LessP (SCM x, SCM y) ;
extern SCM SCM_Float_EqnP (SCM x, SCM y) ;
extern SCM SCM_Read_Char (SCM stream) ;
extern SCM SCM_Peek_Char (SCM stream) ;
extern SCM SCM_String_Length (SCM string) ;
extern SCM SCM_String_Ref (SCM string, SCM index) ;
extern SCM SCM_String_Set (SCM string, SCM index, SCM character) ;
extern SCM SCM_Not (SCM x) ;
extern SCM SCM_Call_CC (SCM f) ;
extern SCM SCM_Bind_DE (SCM key, SCM value, SCM thunk) ;
extern SCM SCM_Assoc_DE (SCM key, SCM success) ;

extern SCM SCM_T ;
extern SCM SCM_F ;
extern SCM SCM_NIL ;
extern SCM SCM_STDIN ;
extern SCM SCM_STDOUT ;
extern SCM SCM_STDERR ;
extern SCM SCM_PAIRP ;
extern SCM SCM_NULLP ;
extern SCM SCM_INTEGERP ;
extern SCM SCM_SYMBOLP ;
extern SCM SCM_STRINGP ;
extern SCM SCM_CHARP ;
extern SCM SCM_ENDOFFILEP ;
extern SCM SCM_PROCEDUREP ;
extern SCM SCM_EQP ;
extern SCM SCM_CONS ;
extern SCM SCM_CAR ;
extern SCM SCM_CDR ;
extern SCM SCM_SETCAR ;
extern SCM SCM_SETCDR ;
extern SCM SCM_LIST ;
extern SCM SCM_PLUS ;
extern SCM SCM_MINUS ;
extern SCM SCM_TIMES ;
extern SCM SCM_QUOTIENT ;
extern SCM SCM_REMAINDER ;
extern SCM SCM_GTP ;
extern SCM SCM_GEP ;
extern SCM SCM_LTP ;
extern SCM SCM_LEP ;
extern SCM SCM_EQNP ;
extern SCM SCM_BASICWRITE ;
extern SCM SCM_INT2CHAR ;
extern SCM SCM_CHAR2INT ;
extern SCM SCM_SYMBOL2STRING ;
extern SCM SCM_MAKE_SYMBOL ;
extern SCM SCM_EXIT ;
extern SCM SCM_GETUNIVERSALTIME ;
extern SCM SCM_FLOAT_PLUS ;
extern SCM SCM_FLOAT_MINUS ;
extern SCM SCM_FLOAT_TIMES ;
extern SCM SCM_FLOAT_DIVIDE ;
extern SCM SCM_FLOAT_GTP ;
extern SCM SCM_FLOAT_LTP ;
extern SCM SCM_FLOAT_GEP ;
extern SCM SCM_FLOAT_LEP ;
extern SCM SCM_FLOAT_EQNP ;
extern SCM SCM_READ_CHAR ;
extern SCM SCM_PEEK_CHAR ;
extern SCM SCM_STRING_LENGTH ;
extern SCM SCM_STRING_REF ;
extern SCM SCM_STRING_SET ;
extern SCM SCM_NOT ;
extern SCM SCM_CALL_CC ;
extern SCM SCM_APPLY ;
extern SCM SCM_BIND_DE;
extern SCM SCM_ASSOC_DE ;

/* Other resources */
extern SCM SCM_error (unsigned int code) ;
extern SCM SCM_report_error(unsigned int code, char *file, unsigned int line);

/* Run-time utilities.
 * This initializes the heap as well as all the predefined objects.
 * The first function is independent of VTS style while the second
 * deals with continuation-specific resources. 				*/

extern void SCM_initialize (int argc, char *argv[]) ;
extern void SCM_report_usage () ;

/* The generic function callers.
 * Fixed arity functions may not have more than 4 variables:
 *    SCM foo (SCM self,x,y,z,t)
 * Otherwise they are represented by SCM foo (int num, SCM args[]).
 * Nary functions have no more than 4 mandatory variables:
 *    SCM bar (SCM x,y, int number, SCM args[])
 * Otherwise they are represented by SCM bar (int num, SCM args[]).
 *
 * Same for primitives which are closures without environment ie
 * without SELF variable.
 * SCM_first_invokation is the first invoker to begin with. It sets up
 * some global variables.						*/

extern void SCM_first_invokation (SCM thunk) ;
extern SCM SCM_invoke0 (SCM fun) ;
extern SCM SCM_invoke1 (SCM fun, SCM) ;
extern SCM SCM_invoke2 (SCM fun, SCM, SCM) ;
extern SCM SCM_invoke3 (SCM fun, SCM, SCM, SCM) ;
extern SCM SCM_invoke4 (SCM fun, SCM, SCM, SCM, SCM) ;
extern SCM SCM_invokeN (SCM fun, int arity, SCM args[]) ;


/*************************************
Things added for chap6f *************/

extern SCM SCM_The_Environment ;
extern SCM SCM_The_Null_Environment ;
extern SCM SCM_The_Dynamic_Environment ;
extern SCM SCM_The_Null_Frame ;

extern SCM SCM_make_function (SCM (*Cfunction)(), SCM environment) ;
extern SCM SCM_make_frame (SIZE_T size) ;
extern SCM SCM_call (SCM fun, SCM frame) ;

extern SCM SCM_Bind_DE (SCM key, SCM value, SCM thunk) ;
extern SCM SCM_Assoc_DE (SCM key, SCM success) ;

extern SCM SCM_List (SCM frame, SCM nullenv) ;

extern SCM SCM_Apply (SCM frame, SCM nullenv) ;

extern SCM SCM_listify (SCM frame, SIZE_T arity) ;

#endif /* RTLIB_SCHEME_H */
